src/Windows/Avalonia.Win32/DataObject.cs

```diff
@@ -8,22 +8,50 @@
 using System.Runtime.InteropServices.ComTypes;
 using System.Runtime.Serialization.Formatters.Binary;
 using Avalonia.Input;
+using Avalonia.MicroCom;
 using Avalonia.Win32.Interop;
+
+using FORMATETC = Avalonia.Win32.Interop.FORMATETC;
 using IDataObject = Avalonia.Input.IDataObject;
 
 namespace Avalonia.Win32
 {
-    class DataObject : IDataObject, IOleDataObject
+    internal static class DataObjectEx
+    {
+        public static unsafe IDataObject GetAvaloniaObject(this Win32Com.IDataObject pDataObj)
+        {
+            switch (pDataObj)
+            {
+                case null:
+                    throw new ArgumentNullException(nameof(pDataObj));
+                case DataObject avnObj:
+                    return avnObj;
+                case OleDataObject oleAvnObj:
+                    return oleAvnObj;
+                default:
+                    var guid = MicroComRuntime.GetGuidFor(typeof(Win32Com.IAvnDataObject));
+                    var rv = ((MicroComProxyBase)pDataObj).QueryInterface(guid, out var ppv);
```

Comment:
QueryReference calls AddRef internally, consider either using MicroComRuntime.QueryInterface<T> to obtain a disposable reference, or adding unwrapping check to MicroComRuntime itself

Reply:
Does it call AddRef on native side? Because this QueryInterface overload only calls ```LocalInterop.CalliStdCallint (PPV, guid, ref, (*PPV)[0])```

Reply:
I added TryQueryInterface

Reply:
https://github.com/AvaloniaUI/Avalonia/tree/feature/microcom-get-managed-reference - merge this into your branch and just use TryUnwrapManagedObject

Reply:
Thanks! Done.